### PR TITLE
Add atomic add opcodes

### LINKED ISSUE #598: Missing opcodes
```
Debug level is 1
Dynarec log level is 1
Dynarec for ARMv8, with extension: HALF FAST_MULT EDSP NEON VFPv4 IDIVA PageSize:4096
Box86 with Dynarec v0.2.5 2d7e3877 built on Mar 27 2022 17:08:23
0xeab2382d: Dynarec stopped because of Opcode 0F C7 61 40 EB 18 0F AE 61 40 EB 12 F7 01 04
0x7bc5a436: Dynarec stopped because of Opcode 66 F0 83 07 01 8D 75 E6 8D 5F 02 8D B4 26 00
```
```
4265|0x3d0052: Unimplemented Opcode (05) 65 05 80 05 6B 05 00 00
```

### PR DESCRIPTION
**Pull request by Trass3r**
https://github.com/ptitSeb/box86/issues/598#issuecomment-1138251631

### CODE REVIEW

File: src/dynarec/dynarec_arm_66f0.c [hunk 1]
```diff
@@ -30,14 +30,69 @@ uintptr_t dynarec66F0(dynarec_arm_t* dyn, uintptr_t addr, uintptr_t ip, int nins
...
+                    DMB_ISH();
+                    if((nextop&0xC0)==0xC0) {
+                        if(opcode==0x81) i16 = F16S; else i16 = F8S;
+                        ed = xEAX+(nextop&7);
+                        emit_add16c(dyn, ninst, ed, i16, x3, x14);
+                    } else {
+                        addr = geted(dyn, addr, ninst, nextop, &wback, x2, &fixedaddress, 0, 0, 1, LOCK_LOCK);
+                        if(opcode==0x81) i16 = F16S; else i16 = F8S;
+                        if(!fixedaddress) {
+                            TSTS_IMM8(wback, 0x3);
```
Comment (ptitSeb) [inline src/dynarec/dynarec_arm_66f0.c:62]: This should be `TSTS_IMM8(wback, 0x1);` as this is 2-bytes test alignment, not a 4-bytes one

**Comment by Trass3r** [comment]
I put up the code here rather than as a diff for ease of use.
But as I said I was just shooting in the dark basing the code on the other `case 0xFF` and on `dynarec_arm_66`:
https://github.com/ptitSeb/box86/blob/master/src/dynarec/dynarec_arm_66.c#L418-L433

File: src/dynarec/dynarec_arm_66f0.c [hunk 2]
```diff
@@ -30,14 +30,69 @@ uintptr_t dynarec66F0(dynarec_arm_t* dyn, uintptr_t addr, uintptr_t ip, int nins
...
+                            B_MARK(cNE);
+                        }
+                        if(!fixedaddress || (fixedaddress && !(fixedaddress&1))) {
+                            MARKLOCK;
+                            LDREX(x1, wback);
```
Comment (ptitSeb) [inline src/dynarec/dynarec_arm_66f0.c:67]: This must be `LDREXH(x1, wback);`, it's a 2-bytes read, not a 4-bytes one

File: src/dynarec/dynarec_arm_66f0.c [hunk 3]
```diff
@@ -30,14 +30,69 @@ uintptr_t dynarec66F0(dynarec_arm_t* dyn, uintptr_t addr, uintptr_t ip, int nins
...
+                            emit_add16c(dyn, ninst, x1, i16, x3, x14);
+                            STREX(x3, x1, wback);
```
Comment (ptitSeb) [inline src/dynarec/dynarec_arm_66f0.c:69]: This is `STREXH(x3, x1, wback);`, again a 2-bytes write

File: src/dynarec/dynarec_arm_66f0.c [hunk 4]
```diff
@@ -30,14 +30,69 @@ uintptr_t dynarec66F0(dynarec_arm_t* dyn, uintptr_t addr, uintptr_t ip, int nins
...
+                            CMPS_IMM8(x3, 0);
+                            B_MARKLOCK(cNE);
+                        }
+                        if(!fixedaddress) {
+                            DMB_ISH();
+                            B_NEXT(c__);
+                        }
+                        if(!fixedaddress || (fixedaddress && (fixedaddress&1))) {
+                            MARK;   // unaligned! also, not enough
+                            LDR_IMM9(x1, wback, 0);
```
Comment (ptitSeb) [inline src/dynarec/dynarec_arm_66f0.c:79]: 2 bytes read, with `LDRH_IMM8(x1, wback, 0);`

File: src/dynarec/dynarec_arm_66f0.c [hunk 5]
```diff
@@ -30,14 +30,69 @@ uintptr_t dynarec66F0(dynarec_arm_t* dyn, uintptr_t addr, uintptr_t ip, int nins
...
+                            LDREXB(x14, wback);
+                            BFI(x1, x14, 0, 8); // re-inject
+                            emit_add16c(dyn, ninst, x1, i16, x3, x14);
+                            STREXB(x3, x1, wback);
+                            CMPS_IMM8(x3, 0);
+                            B_MARK(cNE);
+                            STR_IMM9(x1, wback, 0);    // put the whole value
```
Comment (ptitSeb) [inline src/dynarec/dynarec_arm_66f0.c:86]: 2-bytes write with `STRH_IMM8(x1, wback, 0);`

**Review by ptitSeb — CHANGES_REQUESTED** [review]
You need to change all the access: there are 4-bytes ones, while it needs to be 2-bytes (so with H, like Half-word in ARM vocabulary)

File: src/dynarec/dynarec_arm_66f0.c (hunk 5 above)
Comment (Trass3r) [reply src/dynarec/dynarec_arm_66f0.c]: Thanks! Btw these kind of comments are easier to write and also apply if you use the +/- suggestion button above.

**Review by ptitSeb — APPROVED** [review]

**Pull request merged**